acme/http_challenge.go

```diff
@@ -105,17 +34,17 @@ func (s *httpChallenge) startHTTPServer(domain string, token string, keyAuth str
 		// if the domain:port bind failed, fall back to :port bind and try that instead.
 		listener, err = net.Listen("tcp", port)
 		if err != nil {
-			s.end <- err
+			return fmt.Errorf("Could not start HTTP server for challenge -> %v", err)
 		}
 	}
-	// Signal successfull start
-	s.start <- listener
+	defer listener.Close()
 
-	path := "/.well-known/acme-challenge/" + token
+	path := "/.well-known/acme-challenge/" + chlng.Token
 
 	// The handler validates the HOST header and request type.
 	// For validation it then writes the token the server returned with the challenge
-	http.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+	mux := http.NewServeMux()
```

Comment:
Closing the listener (on line 40 above) will release the listener port and cause http.Serve to return (in the goroutine line 58). listener.Close() will not return until the port is freed, so there should be no race on I/O resources. The http.Serve call will eventually return, but since it has no global state now that it has a private ServeMux, this shouldn't cause an issue.